[MBUILDCACHE-73] Allow arbitrary expressions as additional reconcilation properties

pdudits opened the pull request:
Following this checklist to help us incorporate your
contribution quickly and easily:

- [X] Your pull request should address just one issue, without pulling in other changes.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Each commit in the pull request should have a meaningful subject line and body.
  Note that commits might be squashed by a maintainer on merge.
- [ ] Write unit tests that match behavioral changes, where the tests fail if the changes to the runtime are not applied.
  This may not always be possible but is a best-practice.
- [x] Run `mvn verify` to make sure basic checks pass.
  A more thorough check will be performed on your pull request automatically.
- [x] You have run the integration tests successfully (`mvn -Prun-its verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

- [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
- [X] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

----

The change realizes the approach I originally suggested in [MBUILDCACHE-73](https://issues.apache.org/jira/browse/MBUILDCACHE-73). It is now possible to declare plugin's reconcilation property as 

```xml
<plugin goal="deploy" artifactId="maven-deploy-plugin" groupId="org.apache.maven.plugins">
    <reconciles>
        <reconcile propertyName="project.version" expression="${project.version}"/>
     </reconciles>
</plugin>
```

Along the way I unified how the string value for reconciled expression is constructed -- they were different upon saving and restoration before.

Since this relies on new model properties an initial failing test case wasn't really possible.

---

Fixes #268.

pdudits committed:
[MBUILDCACHE-73] Integration test for reconcile expressions

pdudits committed:
[MBUILDCACHE-73] Unify property value conversions in saving and restoration

pdudits committed:
[MBUILDCACHE-73] Allow arbitrary expressions as additional reconcilation properties

erik-meuwese-topicus commented on src/test/projects/reconcile-expressions/.mvn/extensions.xml:
```diff
@@ -0,0 +1,25 @@
+        <groupId>org.apache.maven.extensions</groupId>
+        <artifactId>maven-build-cache-extension</artifactId>
+        <version>${projectVersion}</version>
+    </extension>
```
XML shoud have 2 space indents to minimize diffs when others use the code style formatter https://maven.apache.org/developers/conventions/code.html

erik-meuwese-topicus commented on src/test/projects/reconcile-expressions/pom.xml:
```diff
@@ -0,0 +1,72 @@
+
+    <dependencies>
+    </dependencies>
+
```
XML shoud have 2 space indents to minimize diffs when other developers use the code style formatter https://maven.apache.org/developers/conventions/code.html

pdudits committed:
Apply suggestions from code review

Co-authored-by: Erik Meuwese <Erik.Meuwese@topicus.nl>

pdudits committed:
Add example use to cache-config template

squash! [MBUILDCACHE-73] Allow arbitrary expressions as additional reconcilation properties

elharo reviewed (changes requested):
"arbitrary" expression strikes me as potentially risky. I need to think about this more

pdudits committed:
Move new utility methods to non-public API

pdudits commented:
Are we good to merge this now, @elharo?

elharo commented:
This should be merged by someone familiar with this plugin who can be assigned to the PR.

pdudits commented:
Er, and who would that be? @olamy?
